src/frequenz/client/base/client.py

```diff
@@ -5,14 +5,102 @@
 
 import abc
 import inspect
+import secrets
+import time
+from base64 import urlsafe_b64encode
 from collections.abc import Awaitable, Callable
 from typing import Any, Generic, Self, TypeVar, overload
 
-from grpc.aio import AioRpcError, Channel
+from cryptography.hazmat.primitives.hashes import SHA256
+from cryptography.hazmat.primitives.hmac import HMAC
+from grpc.aio import (
+    AioRpcError,
+    Channel,
+    ClientCallDetails,
+    Metadata,
+    UnaryUnaryCall,
+    UnaryUnaryClientInterceptor,
+)
 
 from .channel import ChannelOptions, parse_grpc_uri
 from .exception import ApiClientError, ClientNotConnected
 
+RequestT = TypeVar("RequestT")
+ResponseT = TypeVar("ResponseT")
+"""The request and response types of a gRPC method."""
+
+
+# UnaryUnaryClientInterceptor is weird from a typing perspective. It is impossible
+# to make both mypy and the runtime happy at the same time. ChatGPT will hallucinate.
+# You have been warned.
+class MetadataHMACInterceptor(UnaryUnaryClientInterceptor):  # type: ignore
+    """An Interceptor that adds HMAC authentication of the metadata fields to a gRPC call."""
```

Comment:
I would say singing here, this is not really for authentication, as it doesn't know anything about who the user is, it only checks the message was not tampered with.
```suggestion
    """An Interceptor that adds HMAC signing of the metadata fields to a gRPC call."""
```

Reply:
I mean, the signature is used for authentication, as the secret is bound to a specific user. Presumably the API key is not meant to be treated as a secret, and therefore can not really be used for authentication.

That said, I will happily rename it to `signing`

Reply:
No, I believe is the other way around. The secret is chosen by the user, so you can't use it for auth, you don't know who the user is from the secret. The API key is tied to a particular user, and it is the one that lets you know who the user actually is, and yes, it is a secret/credential, users shouldn't share API keys.

But just in case, please correct me if I'm wrong @stefan-brus-frequenz.

Reply:
So two things:
- Even with the user choosing the secret it still needs to be known to the server to verify the signature (because we're using a symmetrically keyed mechanism)
- If you're not denying the message authentication when the signature fails to check out, then having an HMAC doesn't make sense. I would argue the HMAC is for authentication, not least because it is literally called "Message Authentication". Yes, you're not _identifying_ which user it is from. But you are definitely _authenticating_ that the message was created by said user. That is what our signature will be used for. Unless you have a different definition of authentication.

Either way, I named it as you requested :smile:

Reply:
Haha, OK, I see your point, what I mean is not **user** authentication, you are checking the authenticity of the **message** itself, not the sender. But I'm not familiar with the details on how this is implemented on the server side, maybe you can also know the originator of the message if the signature is correct, but in that case I'm not sure what we are using the API key for if the signing key is enough. Anyway, probably out of scope for this PR.

Reply:
I would nitpick that that would be user identification, not authentication. But I definitely see your point as well.
I think it can stay as is (which is as you requested). I am happy to use whatever nomenclature the rest of the company decides on. I think the current problem is there is none.